src/frontend/src/optimizer/plan_visitor/apply_visitor.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::error::{ErrorCode, RwError};
+
+use super::{DefaultBehavior, Merge};
+use crate::optimizer::plan_node::{LogicalApply, PlanTreeNodeBinary};
+use crate::optimizer::plan_visitor::PlanVisitor;
+use crate::PlanRef;
+
+pub struct HasMaxOneRowApply();
+
+impl PlanVisitor for HasMaxOneRowApply {
+    type Result = bool;
+
+    type DefaultBehavior = impl DefaultBehavior<Self::Result>;
+
+    fn default_behavior() -> Self::DefaultBehavior {
+        Merge(|a, b| a | b)
+    }
+
+    fn visit_logical_apply(&mut self, plan: &LogicalApply) -> bool {
+        plan.max_one_row() | self.visit(plan.left()) | self.visit(plan.right())
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Default)]
+enum CheckResult {
+    #[default]
+    Ok,
+    CannotBeUnnested,
+    MoreThanOneRow,
+}
+
+impl From<CheckResult> for Result<(), RwError> {
+    fn from(val: CheckResult) -> Self {
+        let msg = match val {
+            CheckResult::Ok => return Ok(()),
+            CheckResult::CannotBeUnnested => "Subquery can not be unnested.",
+            CheckResult::MoreThanOneRow => "Scalar subquery might produce more than one row.",
+        };
+
+        Err(ErrorCode::InternalError(msg.to_owned()).into())
+    }
+}
+
+#[derive(Default)]
+pub struct CheckApplyElimination {
+    result: CheckResult,
```

Comment:
nit: Why not simply put a `Result<(), RwError>` here

Reply:
We have weight on different error messages to provide better error reporting. 🥺

https://github.com/risingwavelabs/risingwave/blob/32b0c583a74855fd6fccd20c997ae3ad549956b6/src/frontend/src/optimizer/plan_visitor/apply_visitor.rs#L68-L70